# Code of the paper 



- **Benchmark**
  - Cell integration
  - Gene imputation
    
- **Universal cell types**
- **Universal tissue regions**
- **Universal gene embedding**
  - Imputation
  - Targeted gene panel selection
- **Cell-cell interaction**
- **Reference mapping**
  - Benchmark
  - Query dataset
    
- **We also provide a [data portal](https://www.spatial-atlas.net/FuseMap/) of the molecular common coordinate framework (molCCF).**
  - Users can interactively explore universal molecular cell types, molecular tissue regions, and imputed gene expression of 434 sections in the molCCF.


<br>

# Acknowledgement
#### Data source
Publicly available datasets are used in this study (summarized in Supplementary Table 1). The spatial transcriptomics atlases of mouse brain are available in https://singlecell.broadinstitute.org/single_cell/study/SCP1830 (Atlas 1) , https://doi.brainimagelibrary.org/doi/10.35077/g.610 (Atlas 2), https://doi.brainimagelibrary.org/doi/10.35077/act-bag (Atlas 3), https://info.vizgen.com/mouse-brain-map (Atlas 4), http://mousebrain.org/adult/ (Atlas 5), https://db.cngb.org/search/project/CNP0001543/ (Atlas 6), https://www.braincelldata.org/ (Atlas 7). The PANTHER classification of complete protein-coding gene families in mouse genomes is available in http://data.pantherdb.org/ftp/sequence_classifications/current_release/PANTHER_Sequence_Classification_files/PTHR18.0_mouse. The scRNA-seq dataset is available in https://mousebrain.org/. The STARmap PLUS query dataset is available in https://zenodo.org/records/8041114. The MERFISH query dataset is available in https://doi.brainimagelibrary.org/doi/10.35077/g.21. The Stereo-seq query dataset is available in https://github.com/JinmiaoChenLab/SEDR_analyses/tree/master/data. The Slide-seqV2 data can be downloaded from https://singlecell.broadinstitute.org/single_cell/study/SCP815/.

#### Software
The following packages and software were used in the data analysis. MATLAB R2019b and Python 3.6. ImageJ 1.51, R 4.0.4, RStudio 1.4.1106, Jupyter Notebook 6.0.3, Anaconda 2-2-.02, h5py 3.1.0, hdf5 1.10.4, matplotlib 3.1.3, seaborn 0.11.0, scanpy 1.6.0, numpy 1.19.4, scipy 1.6.3, pandas 1.2.3, scikit-learn 0.22, umap-learn0.4.3, pip 21.0.1, numba 0.51.2, tifffile 2020.10.1, scikit-image 0.18.1, anndata 0.8.0, itertools 8.0.0, diffxpy 0.7.4, and gprofiler2 package v0.1.9. Other detailed information of benchmark packages and software are in the ‘Benchmark FuseMap performance’ and ‘Benchmark annotation transfer performance’ section of the paper.

